RNA_folding.py

```diff
@@ -0,0 +1,319 @@
+    nx.draw_networkx_edges(G, pos, edgelist=stem_edges, width=4.5, alpha=0.8, edge_color='tab:pink')
+
+    plt.savefig(fig_name + '.png')
+
+
+def build_cqm(stem_dict, min_stem=3):
+    """ Creates a Constrained Binary Model to optimize most likely stems from a dictionary of possible stems.
+
+    Args:
+        stem_dict: Dictionary with maximal stems as keys and list of weakly contained substems as values.
+        min_stem: Integer minimum number of nucleotides between two bonding nucleotides.
+
+    Returns: Constrained Binary Model.
+    """
+
+    # Create linear coefficients the prioritize inclusion of long stems.
+    linear_coeffs = {stem: -1 * (stem[1] - stem[0] + 1) ** 2 for sublist in stem_dict.values() for stem in sublist}
+
+    # Create constraints for overlapping and and substem containment.
+    quadratic_coeffs = pseudoknot_terms(stem_dict)
+
+    bqm = dimod.BinaryQuadraticModel(linear_coeffs, quadratic_coeffs, 'BINARY')
+
+    cqm = dimod.ConstrainedQuadraticModel()
+    cqm.set_objective(bqm)
+
+    # Add constraint disallowing overlapping substsems included in same maximal stem.
+    for stem, substems in stem_dict.items():
+        if len(substems) > 1:
+            # Add the variable for all zeros case in one-hot constraint
+            zeros = 'Null:' + str(stem)
+            cqm.add_variable(zeros, 'BINARY')
+            cqm.add_discrete(substems + [zeros], stem)
+
+    overlaps = []
+    for stem1, stem2 in combinations(stem_dict.keys(), 2):
+        # Check maximal stems first.
+        if check_overlap(stem1, stem2):
+            # If maximal stems overlap, compare list of smaller stems.
+            for stem_pair in product(stem_dict[stem1], stem_dict[stem2]):
+                if check_overlap(stem_pair[0], stem_pair[1]):
+                    cqm.add_constraint(dimod.quicksum([dimod.Binary(stem) for stem in stem_pair]) <= 1)
+
+    # Add constraint disallowing other overlapping stems.
+    # for stem_pairs in find_overlaps(stem_dict):
+    #     cqm.add_constraint(dimod.quicksum([dimod.Binary(stem) for stem in stem_pairs]) <= 1)
+
+    return cqm
+
+
+def process_cqm_solution(sample_set, verbose=True):
+    """ Processes samples from solution and prints relevant information.
+
+    Prints information about the best feasible solution and returns a list of stems contained in solution.
+    Returns solution as a list of stems rather than a binary string.
+
+    Args:
+        sample_set: dimod sampleset element.
+        verbose: Boolean indicating if function should print additional information.
+
+    Returns: List of stems included in optimal solution, encoded as 4-tuples.
+    """
+
+    infeasible = True
+    for sample in sample_set.data():
+        if sample.is_feasible:
+            solution = sample
+            infeasible = False
+            break
+
+    # Check for feasible solution.
+    if infeasible:
+        print('\nWarning! All solutions infeasible. You may need to try again.')
+        return None
```

Comment:
With latest dimod, this can be replaced with:
```suggestion
    feasible_samples = sample_set.filter(lambda s: s.is_feasible)
    if not feasible_samples:
        print('\nWarning! All solutions infeasible. You may need to try again.')
        return None
    best = feasible.first
```

Then use `best.energy` and `best.sample` in the later code.

Reply:
Nice.